core/corerepo/pinning.go

```diff
@@ -41,10 +41,7 @@ func Pin(n *core.IpfsNode, ctx context.Context, paths []string, recursive bool)
 			return nil, err
 		}
 
-		ctx, cancel := context.WithCancel(ctx)
-		defer cancel()
```

Comment:
no, this should be here, to ensure all derived contexts under `n.Pinning.Pin` cancel out after this function exits. though there is a bug, this context should be cancelled after the Pin exits, not deferred, as the defer won't do anything until the whole function exists:

```go
pCtx, cancel := context.WithCancel(ctx) // use different name to avoid problems
err = n.Pinning.Pin(pCtx, dagnode, recursive)
cancel()
if err !=  nil {
```